EIPS/eip-6617.md

```diff
@@ -14,18 +14,69 @@ created: 2023-02-27
 
 This EIP offers a standard for building a bit-based permission and role system. Each permission is represented by a single bit. By using an `uint256`, up to $256$ permissions and $2^{256}$ roles can be defined. We are able to specify the importance of each permission based on the order of the bits.
 
-
 ## Specification
 
 The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
 
-*Note* The following specifications use syntax from Solidity `0.8.7` (or above)
+_Note_ The following specifications use syntax from Solidity `0.8.7` (or above)
 
 - Permission and role MUST be defined as an `uint256`
 - Permission MUST be defined as a power of two
 - Permission MUST be unique
 - `0` MUST be used for none permission
 
+```solidity
+pragma solidity ^0.8.7;
+
+/**
+    @title EIP-6617 Bit Based Permission
+    @dev See https://eips.ethereum.org/EIPS/eip-6617
+*/
+library EIP6617 {
```

Comment:
Hm, so this poses another problem: an EIP should define a standard that needs to work between multiple different implementations. Generally a Solidity library is _not_ such a standard (because it is a single implementation.)

For this particular proposal, if you want to standardize bitfields for permissions, you might want to do something like:

```solidity
interface IPermissions {
    /// Returns a URL (preferably IPFS or data:...) with JSON metadata describing the permission.
    function permissionMetadata(uint8 perm) external returns (string);

    /// Returns `true` if `actor` has the given permission.
    function hasPermission(address actor, uint8 perm) external returns (bool);
}
```

I'm no Solidity expert though, and haven't compiled the above, so YMMV.

This way the contract is free to implement permissions however they like (likely a bitfield), and there's a reason to standardize: human-readable metadata.

Reply:
@SamWilsn I think, it's a part of its sister proposal EIP-6366 https://github.com/orochi-network/EIPs/blob/master/EIPS/eip-6366.md?plain=1#L144-L166

Reply:
@vdusart Do you think we should move the metadata here?

Reply:
Hi @SamWilsn,
I don't really see why metadata should be in this lib. 😢
Here the library is not used to save permissions (or informations about them) but only to perform actions on them.
I think like @chiro-hiro that the metadatas should be in the eip 6366.😁

Reply:
> ```solidity
> function permissionsOf(address user) external returns (uint256);
> ```

This function alone isn't enough for an EIP. Just... do that in your contract if you want to? Let's break it down a bit, and explore.

Who are the consumers of this proposal? Generally there are three families of consumers of an interface: other contracts, dapps, and wallets.

I'm imagining one contract wants to know about the permissions of some third address. I'm assuming this is the only use case for contracts because if a contract wants to know its own permissions, it can generally just call the function and get rejected or not. I can see this is a valid use case. We're talking about a similar thing over in [ERC-6454's discussion thread](https://ethereum-magicians.org/t/minimalistic-transferable-interface/12517).

While I believe this is a valid use case, I don't think this interface solves it adequately. With just a bitfield, a contract designed to work with multiple ERC-6617 contracts can't know what those permission bits do. You can't be generic over a permissions bitfield without application specific knowledge. If you need application specific knowledge, it isn't a good fit for an EIP.

Similarly here, your dapp/wallet needs to be built to know what the permission bits mean. What good is a user interface that has 256 checks with no labels?

For a rather contrived example, let's say I create a contract and use `1` as the admin privilege, and Panda creates a contract using `2^255`. Both are sane choices. How will any generic interface display useful information about both of these contracts? They'll need application specific information, which again, isn't great for an EIP.

If this proposal created a list of well-known permissions and their bits, or a way to discover information about those permission bits, then I'd be much more amenable to allowing this proposal. As it stands, you've got a named 256-bit integer and I don't believe that's enough to stand on its own.

Reply:
Hey @SamWilsn, @Pandapip1 
To be honest, I'm a little confused and having trouble understanding what you want us to do. To provide some background, earlier this year @chiro-hiro and I proposed an EIP to handle permissions in a different way than the current methods ([173](https://eips.ethereum.org/EIPS/eip-173), [5982](https://eips.ethereum.org/EIPS/eip-5982)).

@Pandapip1 suggested that we split our proposal into two separate EIPs, as this would be more appropriate (see post [here](https://github.com/ethereum/EIPs/pull/6366#pullrequestreview-1294760359)). So we decided to create a first EIP that focuses on bitwise operations (this one), while the second EIP ([6366](https://eips.ethereum.org/EIPS/eip-6366)) handles permissions (transfer, hasPermission, etc.). These two EIPs, 6617 and 6366, work in tandem and the function you are looking for is included in the second EIP (in the core interface [here](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-6366.md#core-interface)).

To answer your comment, @SamWilsn, our goal with these 2 EIPs was to create a solution to manage permissions and roles in an efficient manner. We also wanted to enable adding metadata to give names and descriptions to permissions. All this is possible with EIP 6366.

However, we decided to separate the complex management of bitwise operations into a separate EIP (the 6617).

Reply:
@vdusart You're pretty much getting caught up in a disagreement between @Pandapip1 and myself. ERC-6366 certainly meets all of my requirements to be an EIP!

I sincerely hope I'm not coming off too negatively. I don't mean any of this in a harsh way; I very much appreciate your group's contribution here!

That said, you have two options to satisfy my concerns as an EIP editor: withdraw ERC-6617, or expand ERC-6617.

Withdrawing ERC-6117 is easy, but possibly not the right answer.

Expanding ERC-6617 is harder, but I don't think it'd be prohibitively so. You really just need to add enough "meat" to the specification for it to be useful without ERC-6366. There are plenty of opportunities to do so. The biggest one, to me, is defining some way to figure out what each permission bit means so that they can be displayed in a dapp or wallet. You have `IEIP6366Meta` in ERC-6366; perhaps that belongs here instead? Or maybe something like EIP-712's [`domainSeparator`](https://eips.ethereum.org/EIPS/eip-712#definition-of-domainseparator)? That way you can tell if the permissions bits on one contract have the same meaning on another contract.

I'm not an expert on this permissions stuff. I'm completely open to being convinced that we absolutely need a specification for a bitfield. I just haven't been convinced of that yet.

Reply:
No worries @SamWilsn , this is the kind of situation that can happen when working in a group. Anyway, I didn't take your comment the wrong way, I was just a bit confused, but thanks for helping me improving it !

Let's analyze the two options you gave me:

### Option 1: Withdraw EIP-6617
In this case, we will have to move its content to EIP-6636. Indeed, the contents of EIP-6617 must find a place somewhere, and this will result in a larger EIP-6636 (which did not please @Pandapip1 too much).

### Option 2: Add content to EIP-6617
Actually, in my view, the EIP-6617 is only used to "hide" the bitwise operations required for the EIP-6636 to function. These operations may not be obvious to everyone, and having a library to handle them without worrying about the details seemed like a good idea to me.
Therefore, adding `IEIP6366Meta` here doesn't seem consistent to me with keeping this library only for handling binary operations, and having "actual" permissions handling directly in EIP-6636.

### Option 3
I have a proposal for an option 3, maybe renaming this EIP could reduce confusion about its usefulness. What do you think?

PS: I apologize for the delay in responding. It must be painful to have to dive back into a topic several weeks later. I will be more reactive!

Reply:
@vdusart I think, I'm prefer option 2 here is why:

- EIP-6617 is about "Bit based permission", the bit itself doesn't have any mean. So `IEIP6366Meta` can help to describe the permission better.
- 
We should consider `IEIP6366Meta` is a must or an optional interface.